File: Server/src/main/content/region/misthalin/quest/free/dragonslayer/npc/WormbrainNPC.java
package content.region.misthalin.quest.free.dragonslayer.npc;

import content.region.misthalin.quest.free.dragonslayer.DragonSlayer;
import core.game.node.entity.Entity;
import core.game.node.entity.combat.CombatStyle;
import core.game.node.entity.npc.AbstractNPC;
import core.game.node.entity.player.Player;
import core.game.node.item.GroundItemManager;
import core.game.world.map.Location;

/**
 * The Wormbrain npc.
 */
public final class WormbrainNPC extends AbstractNPC {

    private static final int[] ID = {745};

    /**
     * Instantiates a new Wormbrain npc.
     */
    public WormbrainNPC() {
        super(0, null);
    }

    private WormbrainNPC(int id, Location location) {
        super(id, location);
    }

    @Override
    public AbstractNPC construct(int id, Location location, Object... objects) {
        return new WormbrainNPC(id, location);
    }

    @Override
    public void finalizeDeath(final Entity killer) {
        super.finalizeDeath(killer);
        if (killer instanceof Player) {
            if (((Player) killer).getQuestRepository().getQuest("Dragon Slayer").getStage(killer.asPlayer()) == 20 && !((Player) killer).getInventory().containsItem(DragonSlayer.WORMBRAIN_PIECE) && !((Player) killer).getBank().containsItem(DragonSlayer.WORMBRAIN_PIECE)) {
                GroundItemManager.create(DragonSlayer.WORMBRAIN_PIECE, getLocation(), ((Player) killer));
                ((Player) killer).getPacketDispatch().sendMessage("Wormbrain drops a map piece on the floor.");
            }
        }
    }

    @Override
    public boolean isAttackable(Entity entity, CombatStyle style, boolean message) {
        if (entity instanceof Player) {
            final Player player = (Player) entity;
            if (player.getQuestRepository().getQuest("Dragon Slayer").getStage(player) != 20) {
                if (message) {
                    player.getPacketDispatch().sendMessage("The goblin is already in prison. You have no reason to attack him.");
                }
                return false;
            }
        }
        return super.isAttackable(entity, style, message);
    }

    @Override
    public int[] getIds() {
        return ID;
    }

}

File: Server/src/main/content/region/misthalin/quest/free/dragonslayer/npc/WormbrainNPC.kt
package content.region.misthalin.quest.free.dragonslayer.npc

import content.region.misthalin.quest.free.dragonslayer.DragonSlayer
import core.api.consts.NPCs
import core.game.node.entity.Entity
import core.game.node.entity.combat.CombatStyle
import core.game.node.entity.npc.AbstractNPC
import core.game.node.entity.player.Player
import core.game.node.item.GroundItemManager
import core.game.world.map.Location

/**
 * Wormbrain NPC.
 */
class WormbrainNPC : AbstractNPC {

    constructor() : super(0, null)

    private constructor(id: Int, location: Location) : super(id, location)

    override fun construct(id: Int, location: Location, vararg objects: Any): AbstractNPC {
        return WormbrainNPC(id, location)
    }

    override fun finalizeDeath(killer: Entity) {
        super.finalizeDeath(killer)
        if (killer is Player) {
            if (killer.getQuestRepository().getQuest("Dragon Slayer").getStage(killer.asPlayer()) == 20 && !killer.inventory.containsItem(DragonSlayer.WORMBRAIN_PIECE) && !killer.bank.containsItem(DragonSlayer.WORMBRAIN_PIECE)) {
                GroundItemManager.create(DragonSlayer.WORMBRAIN_PIECE, getLocation(), killer)
                killer.packetDispatch.sendMessage("Wormbrain drops a map piece on the floor.")
            }
        }
    }

    override fun isAttackable(entity: Entity, style: CombatStyle, message: Boolean): Boolean {
        if (entity is Player) {
            val player = entity
            if (player.getQuestRepository().getQuest("Dragon Slayer").getStage(player) != 20) {
                if (message) {
                    player.packetDispatch.sendMessage("The goblin is already in prison. You have no reason to attack him.")
                }
                return false
            }
        }
        return super.isAttackable(entity, style, message)
    }

    override fun getIds(): IntArray {
        return ID
    }

    companion object {
        private val ID = intArrayOf(NPCs.WORMBRAIN_745)
    }
}
